ARROW-11477: [R][Doc] Reorganize and improve README and vignette content

ianmcook opened the pull request.

ianmcook commented:
You can see how this renders by looking at it in my fork at https://github.com/ianmcook/arrow/tree/ARROW-11477/r#readme

nealrichardson commented on r/README.md:
````diff
@@ -22,232 +22,225 @@ access to Arrow memory and messages.
 ```
 
-Or
+Or to switch to the latest nightly development version:
````
```suggestion
If you have `arrow` installed and want to switch to the latest nightly development version, you can use the included `install_arrow()` utility function:
```

nealrichardson commented on r/README.md:
```diff
@@ -22,232 +22,225 @@ access to Arrow memory and messages.
+|-----|----------------|-------------------------------------------|----------------------------------------------------|
+| 0   | `Scalar`       | single value and its `DataType`           | `Scalar$create(value, type)`                       |
+| 1   | `Array`        | vector of values and its `DataType`       | `Array$create(vector, type)`                       |
+| 1   | `ChunkedArray` | list of `Array`s with the same `DataType` | `chunked_array(..., type)`                         |
```
For standardization maybe this should use `ChunkedArray$create()`, just so the pattern is clear. Also IDK that "list of `Array`s with the same `DataType`", while accurate on some level, is the right thing to say here. In terms of interface, it behaves like Array, you don't treat it like a list type.

nealrichardson commented on r/README.md:
```diff
@@ -22,232 +22,225 @@ access to Arrow memory and messages.
+| 1   | `Array`        | vector of values and its `DataType`       | `Array$create(vector, type)`                       |
+| 1   | `ChunkedArray` | list of `Array`s with the same `DataType` | `chunked_array(..., type)`                         |
+| 2   | `RecordBatch`  | list of `Array`s with a `Schema`          | `record_batch(...)`                                |
+| 2   | `Table`        | list of `ChunkedArray` with a `Schema`    | `Table$create(...)` or `arrow::read_*()` functions |
```
perhaps?

```suggestion
| 2   | `Table`        | list of `ChunkedArray` with a `Schema`    | `Table$create(...)` or `arrow::read_*(as_data_frame = FALSE)` functions |
```

nealrichardson commented on r/README.md:
```diff
@@ -22,232 +22,225 @@ access to Arrow memory and messages.
-repository correspond to dates, and you will likely want the most recent.
+Arrow defines the following classes for representing 0-dimensional
+(scalar), 1-dimensional (vector), and 2-dimensional (tabular/data
+frame-like) data:
```
I might set this up differently: start with what R has since we can assume users are familiar with that. R has vectors, and data.frame that collects vectors of equal length. Arrow has two in-memory representations for the same data: one, RecordBatch containing Arrays, is closest to the R model. The other, Table containing ChunkedArrays, supports array data that may be sliced into multiple pieces in order to accommodate larger sizes or other optimizations. A Dataset is similar to a Table but does not hold all of the data in memory.

Then get into differences. Arrow has a true Scalar type, though you as an R user won't generally have to deal with them. Arrow has a richer data type system and metadata (discuss Schema, DataType, and the range of data types there are.)

ianmcook replied:
Ok, gotcha. Based on this and your other comment above, I think it makes sense to add a "Why use this package?" section before the "Installation" section that will establish the framing to put the details here in more of a clearer context.

I could also modify the content here to focus chiefly on the `Table` class and move these tables into a separate vignette, perhaps titled "Apache Arrow object hierarchy" (because it's valuable to have this content presented _somewhere_ in a documentation style as opposed to a narrative style, even if it's not here in the readme).

ianmcook replied:
I moved the object hierarchy content to the "Using the Arrow C++ Library in R" vignette and added just two bullets here describing `Table` and `Dataset` which for many purposes will be the only two data structures users really need to know about.

nealrichardson commented on r/README.md:
````diff
@@ -22,232 +22,225 @@ access to Arrow memory and messages.
-  -DARROW_INSTALL_NAME_RPATH=OFF \
-  ..
-```
+-   `ArrowDatum`: inherited by `Scalar`, `Array`, and `ChunkedArray`
````
I wouldn't include this in the README. This is an implementation detail. There are implementation details in vignette("arrow").

nealrichardson commented on r/README.md:
```diff
@@ -22,232 +22,225 @@ access to Arrow memory and messages.
+-   `read_csv_arrow()`: read a comma-separated values (CSV) file
+-   `read_tsv_arrow()`: a tab-separated values (TSV) file
+-   `read_json_arrow()`: read a JSON data file
+-   `read_feather()`: read a file in Feather format (the Apache Arrow
```
I'd lead with parquet and feather

nealrichardson commented on r/README.md:
```diff
@@ -22,232 +22,225 @@ access to Arrow memory and messages.
+
+If you encounter a bug, please file an issue with a minimal reproducible
+example on [the Apache Jira issue
+tracker](https://issues.apache.org/jira/). Choose the project **Apache
```
I think there is a direct link to the arrow JIRA, should be on our website on the contributing page.

ianmcook replied:
I considered this and tried first to use one of these URLs:
- https://issues.apache.org/jira/projects/ARROW/issues/
- https://issues.apache.org/jira/projects/ARROW/summary
but I tried opening these in incognito windows to see how users would see them if they're not already signed into Jira, and they were all quite confusing.

I also tried constructing a direct link to the Create Issue page by using URL parameters after https://issues.apache.org/jira/secure/CreateIssue.jspa but I couldn't get that working either

ianmcook replied:
I guess I'll just switch this to using https://​issues.apache.org/​jira/​projects/​ARROW/​issues because that's used everywhere else, but oof it's a bad experience for folks who aren't familiar with Jira

ianmcook commented on r/README.md:
```diff
@@ -22,232 +22,225 @@ access to Arrow memory and messages.
-don’t need to alter the Arrow C++ library may be able to obtain a
-recent version of the library without building from source. On macOS,
-you may install the C++ library using [Homebrew](https://brew.sh/):
+Arrow defines the following classes for representing metadata:
```
The benefit of keeping it in this order is that there are no forward references; reading from top to bottom, you never come across something that has not already been defined above.

ianmcook commented on r/README.md:
````diff
@@ -22,232 +22,225 @@ access to Arrow memory and messages.
 ```
 
-Conda users can install `arrow` nightlies from our nightlies channel using:
+Conda users can install `arrow` nightly builds with
````
To clarify: do you mean I should move
``` shell
conda install -c arrow-nightlies -c conda-forge --strict-channel-priority r-arrow
```
to above
``` r
arrow::install_arrow(nightly = TRUE)
```
?

nealrichardson replied:
correct. install_arrow is a convenience for both the CRAN and conda methods of installation

jonkeane commented on r/README.md:
````diff
@@ -15,239 +15,227 @@ The `arrow` package exposes an interface to the Arrow C++ library to
-```
+Each of these is defined as an `R6` class in the `arrow` R package and
+corresponds to a class of the same name in the Arrow C++ library. The
+`arrow` package provides a variety of `R6` and S3 methods for
````
Very minor
 
```suggestion
`arrow` package provides a variety of `R6` and `S3` methods for
```

ianmcook replied:
Elsewhere in our docs, S3 is not set in monospace text, but `R6` is. Was that deliberate?

jonkeane commented on r/README.md:
```diff
@@ -15,239 +15,227 @@ The `arrow` package exposes an interface to the Arrow C++ library to
+The `arrow` package provides functions for reading data from several
+common file formats. By default, calling any of these functions returns
+an R data frame. To return an Arrow `Table`, set argument
+`as_data_frame = FALSE`.
```
Would it be too much of a distraction to have a brief aside about why one might want to do this here?

ianmcook replied:
I think the "Why" section I'm working on adding at the top of the README should answer that 👍

jonkeane commented on r/README.md:
```diff
@@ -15,239 +15,227 @@ The `arrow` package exposes an interface to the Arrow C++ library to
-checkout:
+For writing Arrow tabular data structures to files, the `arrow` package
+provides the functions `write_parquet()` and `write_feather()`. These
+functions also accept R data frames.
```
```suggestion
functions also accept R dataframes.
```

ianmcook replied:
In R, it should be either "data frame" or "data.frame"—don't get me started on how it's totally different if we're talking about pandas or Spark...

jonkeane replied:
I would go for `data.frame` personally, "data frame" looks super weird and 1990s to me!

jonkeane commented on r/README.md:
```diff
@@ -15,239 +15,227 @@ The `arrow` package exposes an interface to the Arrow C++ library to
-this by installing LLVM via Homebrew and running the script as
-`CLANG_FORMAT=$(brew --prefix llvm@8)/bin/clang-format ./lint.sh`
+For larger or multi-file datasets, load the data into a `Dataset` as
+described in `vignette("dataset", package = "arrow")`.
```
I wonder if this would be better below the `dplyr` block below? It might be the GH diff UI, but I thought this was semi-distracting from the example we're working with so far here

ianmcook replied:
I think it's worth having here (or possibly in both places) but I'll add an "Or, " at the beginning to make it clearer that it's just another data loading option like the other two listed immediately above.

jonkeane commented on r/README.md:
```diff
@@ -15,239 +15,227 @@ The `arrow` package exposes an interface to the Arrow C++ library to
-  `minio server` process is found running. If you're running MinIO with custom
-  settings, you can set `MINIO_ACCESS_KEY`, `MINIO_SECRET_KEY`, and
-  `MINIO_PORT` to override the defaults.
+The `arrow` package uses lazy evaluation. `result` is an object with
```
```suggestion
The `arrow` package uses lazy evaluation to delay computation until it is requested. `result` is an object with
```

ianmcook replied:
I think I'll use "until the result is required" instead of "until if that is requested" if that's cool with you

jonkeane replied:
Yeah, that works — I had that at first but then went down a path of "but what does 'required' mean and how do we operationalize it?" but I think that's because I have a developer's perspective

jonkeane commented on r/README.md:
```diff
@@ -15,239 +15,227 @@ The `arrow` package exposes an interface to the Arrow C++ library to
+For `dplyr` queries on `Table` and `RecordBatch` objects, if the `arrow`
+package detects an unsupported function within a `dplyr` verb, it
+automatically calls `collect()` to return the data as an R data frame
+before processing that `dplyr` verb.
```
Should we add something here about what happens on unsupported calls with datasets?

jonkeane commented on r/README.md:
```diff
@@ -15,239 +15,227 @@ The `arrow` package exposes an interface to the Arrow C++ library to
+Arrow](https://arrow.apache.org/docs/developers/contributing.html) page
+in the Arrow developer documentation.
+
+We welcome questions and discussion about the `arrow` package. For
```
```suggestion
We welcome questions and discussion about and contributions to the `arrow` package. For
```

(or something similar)

jonkeane reviewed (commented):
a few minor suggestions/comments

ianmcook commented:
> Thanks for doing this. Some initial thoughts. You're covering a lot of the "whats" here, and I encourage you to think next of the "whys". The README should persuade potential users that there's value in the package for them and that they should try it out. I see here that I can read a parquet file, or create a RecordBatch, or use dplyr on that, but why would I want to do that?

I added a new section "What can `arrow` do?" at the top. This effectively answers the question "Why should I use this?" with "Because it can..." and a list of the major capabilities, each in just a few words.

I moved all the details about the metadata and data objects to the "Using the Arrow C++ Library in R" where I think they fit better and are valuable as a reference for more advanced users. Here in the README, I retained only two short bullets describing what `Table` and `Dataset` are for, in terms that are less technical and more intended to explain what they do that R data frames don't.

I expanded the dplyr example into a broader "Usage" section in which the dplyr content is one subsection.

ianmcook commented on r/README.md:
```diff
@@ -11,243 +11,273 @@ data, organized for efficient analytic operations on modern hardware. It
+`vignette("fs", package = "arrow")`
 
-Fix any style issues before committing with
+### Using dplyr with Arrow
```
We discussed the possibility of adding a separate new vignette about using arrow with dplyr. When I started to do that, I found that most of what needed to be said was not about the dplyr backend functionality but about the Arrow objects that you call them on and the Arrow objects they return. So the dplyr vignette was really turning out to be less than half about dplyr, and mostly about the data reading and writing functions and about `Table` and `Dataset` objects. Since that content belongs here in the README, I think it makes sense to slip in the dplyr content here too, at least for now, especially since the new dplyr capabilities are a marquee feature for the release. But if there are strong feelings that the README is not the right place for these 55 lines of dplyr content, I'm happy to discuss alternatives.

ianmcook commented on r/README.md:
```diff
@@ -4,250 +4,283 @@
+-   Access and manipulate Arrow objects through **low-level bindings**
+    to the C++ library
+-   Provide a **toolkit for building connectors** to other applications
+    and services that use Arrow
```
This **making words bold** thing is kinda cheesy, I know, but it stops this from just looking like a big wall of text.

nealrichardson commented on r/README.md:
```diff
@@ -4,250 +4,283 @@
+`NOT_CRAN=true`. See `vignette("install", package = "arrow")` for
+details.
 
 ## Installing a development version
```
Maybe delete this heading?

ianmcook replied:
I think it's worthwhile to set this apart from the release install info. I'll try changing this and the other to subheadings under an "Installation" heading

nealrichardson reviewed (commented):
Initial round of comments, will keep reading and commenting

ianmcook commented on r/README.md:
```diff
@@ -4,250 +4,283 @@
+    `write_parquet()`), an efficient and widely used columnar format
+-   Read and write **Feather files** (`read_feather()`,
+    `write_feather()`), a format optimized for speed and
+    interoperability
```
Agreed—I cut the words here down to the bone to keep each bullet point on one line when rendered in GitHub

ianmcook replied:
I added some more detail on Parquet further down. I think for now it would be best not to add more links or detail in this section.

jonkeane commented on r/README.md:
```diff
@@ -4,250 +4,283 @@
+tracker](https://issues.apache.org/jira/projects/ARROW/issues). Create
+an account or log in, then click **Create** to file an issue. Select the
+project **Apache Arrow (ARROW)**, select the component **R**, and begin
+the issue summary with **\[R\]** followed by a space. For more
```
On my rendering, the escape of `\[` isn't actually working.

I can't quite tell from these issues if there's a better work around or if this is resolved in the development version of pkgdown (they have switched from markdown_github to gfm) 

https://github.com/rstudio/rmarkdown/issues/1860
https://github.com/r-lib/pkgdown/pull/1193
https://github.com/r-lib/pkgdown/pull/1195

ianmcook replied:
Huh, weird, thanks. Easy dodge: I'll just make it `monospace` so no escapes are needed.

jonkeane replied:
Yeah, I tried the "obvious" `\\[` `\\\[` and nothing worked easily without monospacing it

jonkeane commented on r/vignettes/dataset.Rmd:
```diff
@@ -322,9 +330,9 @@ by calling `write_dataset()` on it:
+Next, let's imagine that the `payment_type` column is something we often filter on,
 so we want to partition the data by that variable. By doing so we ensure that a filter like
-`payment_type == 3` will touch only a subset of files where payment_type is always 3.
+`payment_type == 3` will touch only a subset of files where `payment_type `is always 3.
```
You didn't touch the example below that does 

```
ds %>%
  filter(payment_type == 3) %>%
  write_dataset("nyc-taxi/feather", format = "feather")
```

But that is no longer valid since we don't autocast numerics to strings anymore. The `group_by`/`write_dataset` will all work, but the query with a filter must have the quotes around it now

ianmcook replied:
I fixed this (sort of) in 88edc2d. It's an imperfect fix but I think it's about the best we can do without making major changes

jonkeane commented:
I’ve read through the readme + dataset vignettes and these are both great and improvements over what we had. I'm torn about this next suggestion so feel free to ignore it, but would it be helpful (or more distracting?) to have one-line examples in that first group of bullets in the readme to show how easy each (or most) of those actions are? They would be kind of floating without much context, but it might help get people in and excited to see that it really is just `read_parquet("big_file.parquet")` to get started with parquets.

nealrichardson commented on r/README.md:
```diff
@@ -4,250 +4,284 @@
+access to the Arrow C++ library API and higher-level access through a
+`dplyr` backend and familiar R functions.
+
+## What can the `arrow` package do?
```
```suggestion
What can the `arrow` package do?
```

ianmcook replied:
I think it looks much better with this set as a heading

nealrichardson commented on r/README.md:
```diff
@@ -4,250 +4,284 @@
+uses are:
 
-To enable optional features including S3 support, an alternative memory allocator, and additional compression libraries, add some or all of these flags:
+-   Reading and writing data files
```
So what? Isn't it pretty basic and universal to be able to read and write files?

I know you're trying to be concise, but maybe it would be good to note reasons you'd want to use `arrow` to read your files, such as (1) supports different file formats; (2) fast; (3) can read and write to cloud storage.

nealrichardson commented on r/README.md:
```diff
@@ -4,250 +4,284 @@
 
-To enable optional features including S3 support, an alternative memory allocator, and additional compression libraries, add some or all of these flags:
+-   Reading and writing data files
+-   Manipulating Arrow data with `dplyr` verbs
```
```suggestion
-   Manipulating bigger-than-memory data with `dplyr` verbs
```

nealrichardson commented on r/README.md:
```diff
@@ -4,250 +4,284 @@
-remotes::install_deps(dependencies = TRUE)
-'
+For example, let’s write the Star Wars characters data that’s included
+in `dplyr` to a Parquet file, then read it back in. First load the
```
Why would I want to do this? Perhaps here is where you put the note/link about the Parquet format.

nealrichardson commented on r/README.md:
````diff
@@ -4,250 +4,284 @@
-generation.
+``` r
+sw <- read_parquet(file_path)
+```
````
It would also be good here to mention (and link to the "arrow" vignette) that R attributes are preserved round-trip, so you can save things like `sf` data frames to Parquet. That seems to be a use case of interest.

nealrichardson commented on r/README.md:
```diff
@@ -4,250 +4,284 @@
 
-    ./lint.sh
+All these functions can read and write files in the local filesystem (by
+passing unqualified paths or `file://` URIs) or files in Amazon S3 (by
```
I wouldn't mention `file://` URIs, I think it's confusing

nealrichardson commented on r/README.md:
```diff
@@ -4,250 +4,284 @@
 
-We use Google C++ style in our C++ code. Check for style errors with
+For reading and writing larger files or multiple files with Arrow
+`Dataset` objects, `arrow` provides the functions `open_dataset()` and
```
I would mention here that this does not require reading all of the data into memory, so you can filter and select data for analysis efficiently

nealrichardson commented on r/README.md:
```diff
@@ -4,250 +4,284 @@
-We use Google C++ style in our C++ code. Check for style errors with
+For reading and writing larger files or multiple files with Arrow
+`Dataset` objects, `arrow` provides the functions `open_dataset()` and
+`write_dataset()`. For examples of these, see
```
Writing datasets is not just about the inverse of reading, it's about partitioning, which lets you speed up your queries significantly

nealrichardson commented on r/README.md:
```diff
@@ -4,250 +4,284 @@
-  `minio server` process is found running. If you're running MinIO with custom
-  settings, you can set `MINIO_ACCESS_KEY`, `MINIO_SECRET_KEY`, and
-  `MINIO_PORT` to override the defaults.
+The `arrow` package uses lazy evaluation to delay computation until the
```
Why is delaying computation useful?

nealrichardson commented on r/README.md:
````diff
@@ -4,250 +4,284 @@
+#> 10 R5-D4                   38.2     70.5
 ```
+
+The `arrow` package works with most `dplyr` verbs except those that
````
This is the part where I think that a full dplyr vignette would help, so you could get into these details and provide more examples. I think the support we have is greater than this paragraph implies (not that the words are wrong, it just focuses on what "limited" support there is rather than enumerating all of the things that are supported).

The NEWS.md now has some more discussion that you could include here.

nealrichardson commented on r/README.md:
```diff
@@ -4,250 +4,284 @@
+`group_by()`). Inside `dplyr` verbs, Arrow offers limited support for
+functions and operators, with broader support expected in upcoming
+releases. For more information about available compute functions, see
+`help("list_compute_functions")`.
```
I don't think we want to link to this on the README. This should be deep in the (hypothetical) vignette. I would expect that most users won't call the arrow_namespaced compute functions but will just use the base/stringr/dplyr-mapped versions. Unfortunately there's not a good place to list those (though you could envision adding a function that could list them), which is why I was thinking a vignette would be necessary to document them.

nealrichardson commented on r/vignettes/arrow.Rmd:
```diff
@@ -72,6 +72,45 @@ to other applications and services that use Arrow. One example is Spark: the
 
+# Object hierarchy
+
+## Metadata objects
```
As I suggested before, I'd swap the order of data and metadata subsections, lead with a mapping of R vector/data.frame to Arrow versions, and then in the metadata section, invert the order of the table rows.

ianmcook replied:
Since this has been moved from the README to a less conspicuous place in the "Using the Arrow C++ Library in R" vignette, I don't think this is as important now. I also believe there are some good reasons to keep it in this order, chiefly the fact that it avoids forward references: reading from top to bottom, you never come across something that has not already been defined above. IMO we should keep this as is for now and plan to refine it for the next release.

nealrichardson replied:
Re: forward references, I disagree that this is a virtue, I think it's more readable to start with the thing you care about and recurse down into the details as needed. But agreed that we don't have to resolve this today.

nealrichardson commented on r/vignettes/dataset.Rmd:
```diff
@@ -277,6 +283,8 @@ avoid scanning because they have no rows where `total_amount > 100`.
 
 ## More dataset options
 
+<!-- ADD SOMETHING HERE ABOUT HOW YOU CAN PASS A SINGLE FILE PATH OR VECTOR OF FILE PATHS TO open_dataset() -->
```
Yes, particularly as it relates to write_dataset(). The key point there is that you could have a CSV file that is too big to read into memory, and you could partition it into a bunch of parquet files that are more manageable in size without having to find some way to first read the file into R.

ianmcook commented on r/README.md:
```diff
@@ -4,250 +4,283 @@
+    interoperability
+-   Open or write **large, multi-file datasets** with a single function
+    call (`open_dataset()`, `write_dataset()`)
+-   Read **large CSV and JSON files** with excellent **speed and
```
I don't really think this is the right place for links; this is mean to be a fairly breezy list of key features, with elaboration provided below. I'd prefer to table this idea for later

ianmcook commented:
> I’ve read through the readme + dataset vignettes and these are both great and improvements over what we had. I'm torn about this next suggestion so feel free to ignore it, but would it be helpful (or more distracting?) to have one-line examples in that first group of bullets in the readme to show how easy each (or most) of those actions are? They would be kind of floating without much context, but it might help get people in and excited to see that it really is just `read_parquet("big_file.parquet")` to get started with parquets.

I think for this version, we should keep the bullets very high-level and easy for folks to quickly read through. Let's table this for later consideration.
